### PR TITLE
raftstore/apply: skip notification for synchronous destroy

### PR DESCRIPTION
**Pull request by gengliqi**
Signed-off-by: Liqi Geng <gengliqiii@gmail.com>

###  What have you changed?

No need to send destroy reply to peerfsm if it will be destroyed synchronously. If not, panic will happen because peerfsm is destroyed twice. 
It rarely happens because a destroy can do synchronously only when an applying snapshot is in the pending state. Also, when applying a snapshot, it is not possible to merge or conf change, the other choices are rare, e.g. peer id in msg is bigger than itself, which needs two conf change.

###  What is the type of the changes?

- Bugfix (a change which fixes an issue)

###  How is the PR tested?

- Integration test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No

###  Does this PR affect `tidb-ansible`?

No

###  Refer to a related PR or issue link (optional)

https://github.com/tikv/tikv/issues/5365

### CODE REVIEW

File: src/raftstore/store/fsm/apply.rs [hunk 1]
```diff
@@ -2766,11 +2769,7 @@ impl PollHandler<ApplyFsm, ControlFsm> for ApplyPoller {
             }
         }
         normal.handle_tasks(&mut self.apply_ctx, &mut self.msg_buf);
-        if normal.delegate.merged {
```
Comment (Connor1996) [inline src/raftstore/store/fsm/apply.rs]: seems it's never updated? legacy code? @BusyJay

Comment (gengliqi) [reply src/raftstore/store/fsm/apply.rs]: Yes. It's legacy code.

File: src/raftstore/store/peer.rs [hunk 2]
```diff
@@ -507,7 +507,12 @@ impl Peer {
             );
             return None;
         }
+        // If initialized is false, it implicitly means applyfsm is not exist now.
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: s/is not/does not/

File: src/raftstore/store/peer.rs [hunk 3]
```diff
@@ -507,7 +507,12 @@ impl Peer {
...
         let initialized = self.get_store().is_initialized();
+        // If async_remove is true, it means peerfsm need to be removed after its
+        // corresponding applyfsm was removed.
+        // If it is false, it means either applyfsm is not exist or there is no task
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: s/is not/does not/

File: src/raftstore/store/peer.rs [hunk 4]
```diff
@@ -507,7 +507,12 @@ impl Peer {
             );
             return None;
         }
+        // If initialized is false, it implicitly means applyfsm is not exist now.
         let initialized = self.get_store().is_initialized();
+        // If async_remove is true, it means peerfsm need to be removed after its
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: s/need/needs/

File: src/raftstore/store/fsm/peer.rs [hunk 5]
```diff
@@ -1337,9 +1337,14 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
 
     fn handle_destroy_peer(&mut self, job: DestroyPeerJob) -> bool {
         if job.initialized {
-            self.ctx
-                .apply_router
-                .schedule_task(job.region_id, ApplyTask::destroy(job.region_id));
+            // When initialized is true and async_remove is false, applyfsm is no need to
```
Comment (BusyJay) [inline src/raftstore/store/fsm/peer.rs]: s/is no need/doesn't need/

**Comment by BusyJay** [comment]
I run your test case without your fix several times, the test always pass.

**Comment by gengliqi** [comment]
> I run your test case without your fix several times, the test always pass.

Sorry, I find that I miss a fail point "apply_pending_snapshot". Now this test fails almost every time.

**Comment by gengliqi** [comment]
PTAL @Connor1996 @BusyJay

**Pull request merged**